File: year2019/go/pkg/coders/intCoderDay13.go
package coders

import (
	"fmt"
)

type gameState struct {
	TotalScore int
	BlockCount int
	paddleX    int
	ballX      int
	x, y       int
}

type ArcadeIntCoder struct {
	IntCoder
}

func (aic *ArcadeIntCoder) PlayArcade(codesInput []int) gameState {
	codes := make(map[int]int, len(codesInput))
	for i, v := range codesInput {
		codes[i] = v
	}
	defer func() {
		aic.idx = 0
		aic.relativeBase = 0
	}()
	state := gameState{}
	outputCount := 0
optLoop:
	for {
		execution := codes[aic.idx]
		switch optCode := execution % 100; optCode {
		case 1:
			codes[aic.getIdxFromMode(codes, execution, 3)] =
				codes[aic.getIdxFromMode(codes, execution, 2)] + codes[aic.getIdxFromMode(codes, execution, 1)]
			aic.idx += 4
		case 2:
			bar := codes[aic.getIdxFromMode(codes, execution, 2)] * codes[aic.getIdxFromMode(codes, execution, 1)]
			newIdx := aic.getIdxFromMode(codes, execution, 3)
			codes[newIdx] = bar
			aic.idx += 4
		case 3:
			var input int
			if state.ballX < state.paddleX {
				input = -1
			}
			if state.ballX > state.paddleX {
				input = 1
			}
			codes[aic.getIdxFromMode(codes, execution, 1)] = input
			aic.idx += 2
		case 4:
			output := codes[aic.getIdxFromMode(codes, execution, 1)]
			state = aic.evaluateOutput(state, outputCount, output)
			outputCount++
			aic.idx += 2
		case 5:
			if codes[aic.getIdxFromMode(codes, execution, 1)] != 0 {
				aic.idx = codes[aic.getIdxFromMode(codes, execution, 2)]
				break
			}
			aic.idx += 3
		case 6:
			if codes[aic.getIdxFromMode(codes, execution, 1)] == 0 {
				aic.idx = codes[aic.getIdxFromMode(codes, execution, 2)]
				break
			}
			aic.idx += 3
		case 7:
			var toAssign int
			if codes[aic.getIdxFromMode(codes, execution, 1)] < codes[aic.getIdxFromMode(codes, execution, 2)] {
				toAssign = 1
			}
			codes[aic.getIdxFromMode(codes, execution, 3)] = toAssign
			aic.idx += 4
		case 8:
			var toAssign int
			if codes[aic.getIdxFromMode(codes, execution, 1)] == codes[aic.getIdxFromMode(codes, execution, 2)] {
				toAssign = 1
			}
			codes[aic.getIdxFromMode(codes, execution, 3)] = toAssign
			aic.idx += 4
		case 9:
			aic.relativeBase += codes[aic.getIdxFromMode(codes, execution, 1)]
			aic.idx += 2
		case 99:
			break optLoop
		default:
			panic(fmt.Sprintf("OptCode not known: %d", optCode))
		}
	}
	return state
}

func (aic *ArcadeIntCoder) evaluateOutput(state gameState, outputCount int, output int) gameState {
	switch outputInstruction := outputCount % 3; outputInstruction {
	case 0:
		state.x = output
	case 1:
		state.y = output
	case 2:
		if state.x == -1 && state.y == 0 {
			state.TotalScore = output
			fmt.Printf("Block count: %d, with total score: %d\n", state.BlockCount, state.TotalScore)
			break
		}
		switch output {
		case 2:
			state.BlockCount++
		case 3:
			state.paddleX = state.x
		case 4:
			state.ballX = state.x
		}
	}
	return state
}

File: year2019/cs/Common/IntCoderDay13.cs
namespace Common
{
    public class GameState
    {
        private int outputCount;
        private long ballX;
        private long paddleX;
        private long x;
        private long y;
        public long TotalScore { get; private set; }
        public long BlockCount { get; private set; }

        internal int GetInput()
        {
            if (ballX < paddleX) return -1;
            if (ballX > paddleX) return 1;
            return 0;
        }

        internal void UpdateFromOutput(long output)
        {
            switch (outputCount % 3)
            {
                case 0:
                    x = output;
                    break;
                case 1:
                    y = output;
                    break;
                case 2:
                    if (x == -1 && y == 0)
                    {
                        TotalScore = output;
                        break;
                    }
                    switch (output) {
                        case 2:
                            BlockCount++;
                            break;
                        case 3:
                            paddleX = x;
                            break;
                        case 4:
                            ballX = x;
                            break;
                    }
                    break;
            }
            outputCount++;
        } 
    }

    public partial class IntCoder
    {
        public GameState PlayArcade(IList<int> codesInput)
        {
            var codes = new DefaultDict<long, long>();
            for (int i = 0; i < codesInput.Count; i++)
            {
                codes.Add(i, codesInput[i]);
            }
            var state = new GameState();
            try
            {
                do
                {
                    var execution = codes[idx];
                    switch (execution % 100)
                    {
                        case 1:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 2)] + codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 4;
                            break;
                        case 2:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 2)] * codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 4;
                            break;
                        case 3:
                            codes[GetIdxFromMode(codes, execution, 1)] = state.GetInput();
                            idx += 2;
                            break;
                        case 4:
                            var output = codes[GetIdxFromMode(codes, execution, 1)];
                            state.UpdateFromOutput(output);
                            idx += 2;
                            break;
                        case 5:
                            if (codes[GetIdxFromMode(codes, execution, 1)] != 0)
                            {
                                idx = codes[GetIdxFromMode(codes, execution, 2)];
                                break;
                            }
                            idx += 3;
                            break;
                        case 6:
                            if (codes[GetIdxFromMode(codes, execution, 1)] == 0)
                            {
                                idx = codes[GetIdxFromMode(codes, execution, 2)];
                                break;
                            }
                            idx += 3;
                            break;
                        case 7:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 1)] < codes[GetIdxFromMode(codes, execution, 2)] ?
                                1 : 0;
                            idx += 4;
                            break;
                        case 8:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 1)] == codes[GetIdxFromMode(codes, execution, 2)] ?
                                1 : 0;
                            idx += 4;
                            break;
                        case 9:
                            relativeBase += codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 2;
                            break;
                        case 99:
                            idx = codes.Count;
                            break;
                        default:
                            throw new Exception($"OptCode not known {execution}");
                    }
                } while (idx < codes.Count);

                return state;
            }
            finally
            {
                idx = 0;
                relativeBase = 0;
            }
        }
    }
}